### PR TITLE
[Feature] Private README.md for organization

### LINKED ISSUE #29503: Support a member-only organisation profile README
### Feature Description

Many thanks to everyone who worked on the development of features `.profile/README.md`! It's great!

When so much has been done for public profiles, I would like to know what you think about private `README.md`?

I won’t tell you why this is needed, because everyone will determine for themselves why they need it. I'll just leave this link to GitHub https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/customizing-your-organizations-profile#adding-a-member-only-organization-profile-readme

Personally, my company and I use this functionality on GitHub and we really want to see something similar in Gitea. But first I would like to discuss the issue with the community

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by changchaishi**
Implemented #29503

Changes:
1. Since individual users and organizations use the same `.profile` repository name for `README.md`, adding the private one, I rename the contexts with `Public` and `Private` inside them.
2. I wanted to load the private profile the same as GitHub, which uses the `view_as` query parameter. Still, the existing query scheme is quite limited due to the integration with the search form and the repository paginator, so my workaround is to forcefully bring augmented query strings of member/public to the tmpl file.
3. The drop-down is adopted from the search box's sort by component, we need to discuss the stylings.

Things lack and need further guidance (update: all done)
1. [x] At the create repository page, the ideal user experience is when the repo name is entered `.profile` the *make repository private* checkbox should be disabled and stay unchecked, whereas `.profile-private`, the checkbox should be disabled and force checked.
2. [x] Follow 1., Tool tips should be added.
3. [x] The translation and CSS for the current drop box and view as public/member feature.

Screenshots:

image

### CODE REVIEW

File: templates/org/home.tmpl [hunk 1]
```diff
@@ -5,8 +5,24 @@
...
+				</div>
+				{{end}}
+				{{if .IsViewerMember}}
+					{{if and .ShowMemberAndTeamTab .PrivateProfileReadme}}
+						<div id="readme_profile" class="markup">{{.PrivateProfileReadme}}</div>
+					{{end}}
+				{{else}}
+					{{if .PublicProfileReadme}}
+						<div id="readme_profile" class="markup">{{.PublicProfileReadme}}</div>
+					{{end}}
```
Comment (wxiaoguang) [inline templates/org/home.tmpl]: It seems not easy as it looks.

The "view as" not only affects the "profile REAMDE", but also affects the repositories listed and other permission-related contents.

**Comment by changchaishi** [comment]
Hi, I listed the behaviors that this PR implemented as below:

- hints to create `.profile` and `.profile-private` repos (if users do not follow the hint, profile may not show)
private_hint

- `view_as` drop down only shows when the user is member in organization and both public and member profiles exist
only_shows_when_2_present

- If there is no profile repo, not showing `overview` tab
no_profiles_no_overview

- When user is a member of org and logged in, and only 1 profile is present, show that profile 
public_only_show_public
private_shows_private

- When user is not logged in, or not a member, only public profile will show if it is present
non_member_can_show

File: templates/org/home.tmpl (hunk 1 above)
Comment (changchaishi) [reply templates/org/home.tmpl]: Hi, any suggestions on how to fix this?

Comment (wxiaoguang) [reply templates/org/home.tmpl]: I think you need to follow the permission system (org team member) to make the content displayed consistent as "view as public/member"

Comment (changchaishi) [reply templates/org/home.tmpl]: Fixed in https://github.com/go-gitea/gitea/pull/32872/commits/4347353d3e3233adb06f9d54a9c22e3eabd3893c

In my PR, the `FindUserProfileReadme` and `RenderOrgHeader` already resolve whether a Doer can access or read the `.profile-private/README.md` file, and the function is propagated to `teams.go` and `members.go` too. For the `home.tmpl` just make sure to show the drop-down of `view as` when it is needed. The latest commit ensures `home.go` can handle the param if only one or none of the profiles exists it shows available profile only and ignore the parameter.

**Comment by wxiaoguang** [comment]
Could you add some tests in `tests/integration/org_*.go` to verify the "view as public" and "view as member" outputs?

If I understand correctly: view as public: private repositories won't be listed.

**Comment by changchaishi** [comment]
> If I understand correctly: view as public: private repositories won't be listed.

In this design `view as` is only to select profiles. The permission system decides whether the doer can access the README.md. Nothing has changed.

So viewing as public does not mean that a member user will not see private repos listed if he/she already has permission.

**Comment by wxiaoguang** [comment]
> > If I understand correctly: view as public: private repositories won't be listed.
> 
> In this design `view as` is only to select profiles. The permission system decides whether the doer can access the README.md. Nothing has changed.

Maybe GitHub also changed their behavior: `You are viewing the README and pinned repositories as a public user.` vs `You are viewing the README and repositories as a public user.` I wouldn't say it is good but it doesn't block since GitHub does so.

While I still think it needs to add some tests in `tests/integration/org_*.go` to verify the "view as public" and "view as member" outputs, to cover the new code's behavior.

**Comment by changchaishi** [comment]
> While I still think it needs to add some tests in `tests/integration/org_*.go` to verify the "view as public" and "view as member" outputs, to cover the new code's behavior.

I'm working on it. Trying to get familiar with the integration tests.

File: templates/repo/create.tmpl [hunk 2]
```diff
@@ -44,6 +44,8 @@
 					<div class="inline required field {{if .Err_RepoName}}error{{end}}">
 						<label for="repo_name">{{ctx.Locale.Tr "repo.repo_name"}}</label>
 						<input id="repo_name" name="repo_name" value="{{.repo_name}}" autofocus required maxlength="100">
+						<span id="repo_name_public_profile_hint" style="display:none" class="help">{{ctx.Locale.Tr "repo.repo_name_public_profile_hint"}}</span>
```
Comment (lunny) [inline templates/repo/create.tmpl]: Use `tw-hidden` class rather than `style="display:none"`.

Comment (changchaishi) [reply templates/repo/create.tmpl]: fixed in https://github.com/go-gitea/gitea/pull/32872/commits/84ef52704a1fab080e120181cc750c2d336d1732

**Comment by changchaishi** [comment]
Hi @wxiaoguang 

> While I still think it needs to add some tests in `tests/integration/org_*.go` to verify the "view as public" and "view as member" outputs, to cover the new code's behavior.

Tests added in https://github.com/go-gitea/gitea/pull/32872/commits/48b4be2e64cef88ffcf02354afab1e8196c36124

**Comment by wxiaoguang** [comment]
> Tests added in [48b4be2](https://github.com/go-gitea/gitea/commit/48b4be2e64cef88ffcf02354afab1e8196c36124)

Thank you very much, will take a look in a few days.

ps: I think we will make some improvements (push into this branch) if you don't mind.

**Comment by wxiaoguang** [comment]
The exiting code base has some problems using "git repo", I have prepared a refactoring change in https://github.com/wxiaoguang/gitea/tree/refactor-reqctx but it hasn't been submitted to a PR.

But this PR also needs to handle these tricks, so I take some code here to make the git repo work correctly.

File: templates/org/home.tmpl [hunk 3]
```diff
@@ -24,6 +24,31 @@
...
+							</span>
+						{{svg "octicon-triangle-down" 14 "dropdown icon"}}
+						<div class="menu">
+							{{/* TODO: does it really need to use CurrentURL with query parameters? Why not construct a new link with clear parameters */}}
+							<a href="{{QueryBuild $.CurrentURL "view_as" "public"}}" class="item {{if not .IsViewerMember}}selected{{end}}">
+								{{svg "octicon-check" 14 (Iif (not .IsViewerMember) "" "tw-invisible")}} {{ctx.Locale.Tr "settings.visibility.public"}}
+							</a>
+							<a href="{{QueryBuild $.CurrentURL "view_as" "member"}}" class="item {{if .IsViewerMember}}selected{{end}}">
+								{{svg "octicon-check" 14 (Iif .IsViewerMember "" "tw-invisible")}}  {{ctx.Locale.Tr "org.members.member"}}
+							</a>
```
Comment (wxiaoguang) [inline templates/org/home.tmpl]: This is copied from your code (use current request's query parameters)

But I am not sure why we should do it, why it needs to keep all the query parameters but not generate a new link with clear parameters?

----

Update: I think `?view_as=public` is enough bc5718d4aa35d4984931e557337284c52eb4d966

**Comment by wxiaoguang** [comment]
@changchaishi I think we need to move the view-as dropdown to the sidebar (just like GitHub) then we could have more spaces for hints and we could add more features to it in the future.

But I have a question here: https://github.com/go-gitea/gitea/pull/32872/files#r1899536547

**Review by wxiaoguang — APPROVED** [review]
Almost done from my side (the remaining question is https://github.com/go-gitea/gitea/pull/32872#discussion_r1899536547, I think `?view_as=public` is enough, bc5718d4aa35d4984931e557337284c52eb4d966)

**Comment by lunny** [comment]
Looks like the dropdown is not in the right place.

image

**Comment by wxiaoguang** [comment]
> Looks like the dropdown is not in the right place.

It is https://github.com/go-gitea/gitea/pull/32872#issuecomment-2565470780 : 

> I think we need to move the view-as dropdown to the sidebar (just like GitHub) then we could have more spaces for hints and we could add more features to it in the future.

File: templates/org/home.tmpl (hunk 3 above)
Comment (changchaishi) [reply templates/org/home.tmpl]: I wanted to make `view_as` `sort` and `language` work at the same time. But I just noticed Github reset query params when view_as is applied.

Comment (wxiaoguang) [reply templates/org/home.tmpl]: Then let's keep it simple, I do not see it is useful to keep other parameters.

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Then we can add "org profile wiki" link easily:  Add profile wiki to org home #32947

**Pull request merged**